modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2020,12 +2020,16 @@ def sort_function(df):
         if len(self.axes[0]) == 0 or len(self.axes[1]) == 0:
             return self.copy()
         # If this df only has one row partition, we don't want to do a shuffle and sort - we can
-        # just do a full-axis sort.
-        if len(self._partitions) == 1:
+        # just do a full-axis sort. Alternatively, if the total amount of data we have is less than
+        # 3x the number of partitions we have, then our new sorting algorithm will try to upsample
+        # the data. In this case, its probably best to do a full axis sort and re-partition, since
+        # our data is probably overpartitioned.
+        if len(self._partitions) == 1 or len(self.index) < 3 * len(self._partitions):
```

Comment:
I wonder why we check total length instead of checking partition sizes skew, say, if we have input data with partitions of lengths `[100, 100, 1, 100, 100]`, won't sorting just break the same way?

Reply:
Not quite - because the problem is a little subtler than just that we're upsampling small partitions. The problem has to do with how we're determining the probability when we sample our partitions. We borrow from an algorithm called TeraSort for sampling and splitting partitioned data into bins with a roughly equal amount of data per bin. The algorithm has us sample with probability `(num_partitions / total_data_length) * log(num_partitions * total_data_length)`. This fraction is greater than 1 until `total_data_length ≈ 3 * num_partitions`, so the error we get is because we're trying to sample with probability greater than 1. In the case where `total_data_length < 3 * num_partitions` we are overpartitioned. Concretely, this would be like having 16 rows and having all 16 partitions. In this case, rather than upsampling and using our algorithm, it makes more semantic sense (and is likely faster - although I haven't tested this) to just make 1 axis partition, and sort all of the data, and then repartition (to hopefully get fewer, more filled, partitions). 

In the example you give above, we don't run into the bug, since we'll try and sample each partition with probability 0.04 - so when we get to the partition with skew, there's a 4% chance it will be in our overall sample, which is fine, since it's just 1 data point, and the algorithm will continue with no errors. 

One thing I just noticed though is that this approach only succeeds if we have 1 column partition - it may be better here to borrow from the previous sort by doing an axis partition on the set of row partitions that have the sort column key(s) and then sorting and reindexing, as this code will break if we are block partitioned with more than one column (although that would imply we have a relatively short and wide df, which I imagine are fairly rare).

Reply:
@RehanSD What status? Do I understand correctly that you are going to change the condition for `if` statement.

Also, at least we will need to add a test.